### PR TITLE
Explicit memory lowering passes.

### PR DESCRIPTION
**Pull request by lkorenc**
Add passes that lower `hl` value categories types (`hl.lvalue` currently) and operations that deal with memory (allocation, load, store) into `ll`.
This is partly extraction from our conversion to `LLVM` dialect, however it is a logical unit that is useful for us as a separate pass in the pipeline.

As a side-effect, `conv::abi` is now expected to run on module that is already modelled with the explicit memory - while nothing prevents us from having two versions, it is more convenient (and fitting) to have this dependency. We can always supply another abi lowering pass into `hl` later.

### CODE REVIEW

File: include/vast/Conversion/ABI/AggregateTypes.hpp [hunk 1]
```diff
@@ -62,6 +62,33 @@ namespace vast::conv::abi
...
+                auto root_type = root->getResultTypes()[0];
+                if (auto ptr_type = mlir::dyn_cast< hl::PointerType >(root_type))
+                    return ptr_type.getElementType();
+                return root_type;
+            }();
+
+            auto def = hl::definition_of(trg_type, module_op);
+            std::size_t idx = 0;
+            for (const auto &[name, type] : def.getFieldsInfo()) {
+                auto ptr_type = hl::PointerType::get(root->getContext(), type);
```
Comment (xlauko) [inline include/vast/Conversion/ABI/AggregateTypes.hpp:84]: Can not you create a type here that is not in DL?

File: include/vast/Dialect/LowLevel/LowLevelOps.td [hunk 2]
```diff
@@ -24,6 +24,68 @@ def LowLevel_StructGEPOp
...
+  let description = [{ VAST array subscript operator }];
+
+  let assemblyFormat = [{
+    $array `at` ` ` `[` $index `:` type($index) `]` attr-dict
+      `:` type($array) `->` type($result)
+  }];
+}
+
+def ArgAlloca
+    : LowLevel_Op< "arg_alloca" >,
```
Comment (xlauko) [inline include/vast/Dialect/LowLevel/LowLevelOps.td:44]: Put commas on the next line

File: include/vast/Dialect/LowLevel/LowLevelOps.td [hunk 3]
```diff
@@ -24,6 +24,68 @@ def LowLevel_StructGEPOp
...
+        attr-dict `:` type($result)
+    }];
+}
+
+def Load
+    : LowLevel_Op< "load" >,
+      Arguments<(ins AnyType:$ptr)>,
+      Results<(outs AnyType:$result)>
+{
+    let summary = "Load.";
```
Comment (xlauko) [inline include/vast/Dialect/LowLevel/LowLevelOps.td:71]: Extend summaries.

File: include/vast/Dialect/LowLevel/LowLevelOps.td [hunk 4]
```diff
@@ -24,6 +24,68 @@ def LowLevel_StructGEPOp
...
+    let summary = "Alloca for a piece of memory. For now does not support array types.";
+
+    let assemblyFormat = [{
+        attr-dict `:` type($result)
+    }];
+}
+
+def Load
+    : LowLevel_Op< "load" >,
+      Arguments<(ins AnyType:$ptr)>,
```
Comment (xlauko) [inline include/vast/Dialect/LowLevel/LowLevelOps.td:68]: These should have some `PointerLike` constraint. Same holds for store.

File: include/vast/Dialect/LowLevel/LowLevelOps.td [hunk 5]
```diff
@@ -24,6 +24,68 @@ def LowLevel_StructGEPOp
...
+}
+
+def Store
+    : LowLevel_Op< "store" >,
+      Arguments<(ins AnyType:$val, AnyType:$ptr)>
+{
+    let summary = "Store.";
+
+    let assemblyFormat = [{
+        operands attr-dict `:` functional-type(operands, results)
```
Comment (xlauko) [inline include/vast/Dialect/LowLevel/LowLevelOps.td:85]: why `functional-type` when it does not have results?

File: lib/vast/Conversion/ABI/EmitABI.cpp [hunk 6]
```diff
@@ -833,16 +826,20 @@ namespace vast
...
+                return true;
             };
 
-            target.addDynamicallyLegalOp< hl::FuncOp >(should_transform);
+            target.markUnknownOpDynamicallyLegal(should_transform);
+            target.addLegalOp< abi::FuncOp >();
 
             mlir::RewritePatternSet patterns(&mctx);
             patterns.add< func_type< hl::FuncOp > >(tc, abi_info_map, mctx);
+            patterns.add< func_type< ll::FuncOp > >(tc, abi_info_map, mctx);
```
Comment (xlauko) [inline lib/vast/Conversion/ABI/EmitABI.cpp:842]: Can not you generalize these patterns for FunctionTypeInterface or something like that?

File: lib/vast/Conversion/ABI/LowerABI.cpp [hunk 7]
```diff
@@ -217,19 +217,14 @@ namespace vast
...
-
-                    VAST_UNREACHABLE("ABI conversion could not make lvalue for {0}", *val);
+                    auto mctx = direct.getContext();
+                    auto type = hl::PointerType::get(
+                        mctx,direct.getOperand(0).getType());
+                    auto memory = state.rewriter.template create< ll::UninitializedVar >(
+                        direct.getLoc(), type);
+                    return state.rewriter.template create< ll::InitializeVar >(
+                        direct.getLoc(), type, memory, direct.getOperand(0));
+                    return val;
```
Comment (xlauko) [inline lib/vast/Conversion/ABI/LowerABI.cpp:227]: unreachable return? O.o

File: lib/vast/Conversion/ABI/LowerABI.cpp [hunk 8]
```diff
@@ -217,19 +217,14 @@ namespace vast
...
-                        if (implicit_cast.getKind() == hl::CastKind::LValueToRValue)
-                        {
-                            return implicit_cast.getOperand().getDefiningOp();
-                        }
-                    }
-
-                    VAST_UNREACHABLE("ABI conversion could not make lvalue for {0}", *val);
+                    auto mctx = direct.getContext();
+                    auto type = hl::PointerType::get(
+                        mctx,direct.getOperand(0).getType());
```
Comment (xlauko) [inline lib/vast/Conversion/ABI/LowerABI.cpp:222]: spacing

File: lib/vast/Conversion/FromHL/Passes.cpp [hunk 9]
```diff
@@ -40,13 +40,25 @@ namespace vast::conv::pipeline {
         return pass(createHLToLLFuncPass);
     }
 
+    pipeline_step_ptr fn_args_to_alloca() {
+        // TODO add dependencies
```
Comment (xlauko) [inline lib/vast/Conversion/FromHL/Passes.cpp:44]: you can add dependencies if you know them or remove comment if none

File: lib/vast/Conversion/FromHL/Passes.cpp [hunk 10]
```diff
@@ -40,13 +40,25 @@ namespace vast::conv::pipeline {
...
+        return pass(createFnArgsToAllocaPass);
+    }
+
+    pipeline_step_ptr lower_value_categories() {
+        // TODO add dependencies
```
Comment (xlauko) [inline lib/vast/Conversion/FromHL/Passes.cpp:49]: Same here

File: lib/vast/Conversion/Generic/LowerValueCategories.cpp [hunk 11]
```diff
@@ -0,0 +1,559 @@
...
+
+                        return builder
+                            .create< mlir::UnrealizedConversionCastOp >(loc, resultType, inputs)
+                            .getResult(0);
+                    }
+                );
+            }
+        };
+
+#define VAST_DEFINE_REWRITE \
```
Comment (xlauko) [inline lib/vast/Conversion/Generic/LowerValueCategories.cpp:101]: Add undef at end.

File: lib/vast/Conversion/Generic/LowerValueCategories.cpp [hunk 12]
```diff
@@ -0,0 +1,559 @@
...
+            return std::is_same_v< Tag, postfix_tag >;
+        }
+
+        template< typename op_t, typename Trg, typename YieldAt >
+        struct unary_in_place : base_pattern< op_t >
+        {
+            using base = base_pattern< op_t >;
+            using base::base;
+
+            logical_result matchAndRewrite(
```
Comment (xlauko) [inline lib/vast/Conversion/Generic/LowerValueCategories.cpp:341]: this is also VAST_DEFINE_REWRITE?

File: lib/vast/Conversion/ToLLVM/Passes.cpp [hunk 13]
```diff
@@ -14,7 +14,8 @@ namespace vast::conv::pipeline {
     pipeline_step_ptr irs_to_llvm() {
         // TODO add dependencies
```
Comment (xlauko) [inline lib/vast/Conversion/ToLLVM/Passes.cpp:15]: Remove comment

File: lib/vast/Conversion/ToLLVM/Passes.cpp [hunk 14]
```diff
@@ -14,7 +14,8 @@ namespace vast::conv::pipeline {
...
         return pass(createIRsToLLVMPass)
-            .depends_on(to_ll);
+            .depends_on(to_ll)
```
Comment (xlauko) [inline lib/vast/Conversion/ToLLVM/Passes.cpp:17]: `depends_on` is variadic

File: lib/vast/Frontend/Pipelines.cpp [hunk 15]
```diff
@@ -54,7 +54,7 @@ namespace vast::cc {
 
         conversion_path default_conversion_path = {
             { target_dialect::high_level, { reduce_high_level } },
-            { target_dialect::std, { standard_types, abi } },
+            { target_dialect::std, { standard_types } },
```
Comment (xlauko) [inline lib/vast/Frontend/Pipelines.cpp:57]: why did you remove abi here?

**Review by xlauko — COMMENTED** [review]
Add tests on new conversions.  All operation kinds should be chacked you stip lvalue categories correctly. Test generation of argument allocas.

I am not sure whether you are not breaking DL on some places as you are creating new types.

File: include/vast/Conversion/ABI/AggregateTypes.hpp (hunk 1 above)
Comment (lkorenc) [reply include/vast/Conversion/ABI/AggregateTypes.hpp:84]: Technically yes, but since all pointers are the same size DL will work as long as one pointer is there. This need more general fix, will give it a look over all our passes once we are on llvm-18.

File: lib/vast/Conversion/Generic/LowerValueCategories.cpp (hunk 11 above)
Comment (lkorenc) [reply lib/vast/Conversion/Generic/LowerValueCategories.cpp:101]: It is inside a `.cpp` so it is not strictly needed but I will add it.

File: lib/vast/Frontend/Pipelines.cpp (hunk 15 above)
Comment (lkorenc) [reply lib/vast/Frontend/Pipelines.cpp:57]: Because it did not work otherwise :P - since now `abi` requires a lot of `to-ll` passes I think it could not be included here?

File: lib/vast/Conversion/ABI/LowerABI.cpp (hunk 8 above)
Comment (lkorenc) [reply lib/vast/Conversion/ABI/LowerABI.cpp:222]: I will run clang-format on the whole `abi` component, but it will be done in a separate PR. Fixed this issue.

**Review by xlauko — APPROVED** [review]

**Review by xlauko — APPROVED** [review]

**Review by xlauko — APPROVED** [review]
